cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py

```diff
@@ -0,0 +1,412 @@
+    return colors
+
+
+# method to draw the identified communities
+def draw_communities_graph(graph, colors, com):
+    color_map = []
+    for node in graph:
+        for i in range(len(com.communities)):
+            if node in com.communities[i]:
+                color_map.append(colors[i])
+    nx.draw(graph, node_color=color_map, with_labels=True)
+    plt.show()
+
+
+# %%
+indp2 = faiss.index_factory(
+    sentence_embeddings2.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings2)  # to be used for inner product computation
+# setting the number of neighbors to consider for graph connectivity
+neighbors2 = 3
+indp2.train(sentence_embeddings2)
+indp2.add(sentence_embeddings2)
+distance_matric2, positions2 = indp2.search(
+    sentence_embeddings2, sentence_embeddings2.shape[0]
+)
+
+# %%
+B = kneighbors_graph(
+    distance_matric2, neighbors2, mode="connectivity", include_self=False
+)
+B.toarray()[0]
+
+# %%
+# load the graph
+G2 = nx.from_numpy_matrix(B, create_using=nx.Graph())
+print(G2)
+# visualize the graph
+nx.draw(G2, with_labels=True)
+
+# %%
+# applying a community algorithm to the graph to identify subgroups
+coms2 = algorithms.leiden(G2)
+draw_communities_graph(G2, generate_colors(coms2), coms2)
+
+# %%
+list(to_use2.comment[coms2.communities[0]])  # Observation on hand washing
+
+# %%
+list(to_use2.comment[coms2.communities[2]])  # belief that the disease exists
+
+# %%
+list(to_use2.comment[coms2.communities[1]])  # belief on wearing of masks
+
+# %%
+list(to_use2.comment[coms2.communities[0]])  # Believe masks are not effective
+
+# %%
+
+# %% [markdown]
+# # Working with the larger dataset
+
+# %%
+indp = faiss.index_factory(
+    sentence_embeddings.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings)  # to be used for inner product computation
+neighbors = 10
+
+# %%
+indp.train(sentence_embeddings)
+indp.add(sentence_embeddings)
+distance_matric, position = indp.search(
+    sentence_embeddings, sentence_embeddings.shape[0]
+)  # create similarity matrix for the data
+
+# %%
+# create a KNN network with n neighbors using the cosine similarity matrix
+A = kneighbors_graph(distance_matric, 5, mode="connectivity", include_self=False)
+A.toarray()[0]
+
+# %%
+# load the graph
+G = nx.from_numpy_matrix(A, create_using=nx.Graph())
+print(G)
+# visualize the graph
+nx.draw(G, with_labels=False)
+
+# %%
+len(G.nodes), len(G.edges)
+
+# %%
+# using leiden community detection algorithm on the dataset
+coms = algorithms.leiden(G)
+draw_communities_graph(G, generate_colors(coms), coms)
+
+# %%
+len(coms.communities)
+
+# %%
+for index in range(len(coms.communities)):
+    print(data.comment[coms.communities[index]].index)
+    print("######################NEW Group####")
+
+# %%
+list(
+    data.iloc[[158, 161, 162, 163, 164, 165, 166, 167, 168, 171]].comment
+)  # belief about mask wearing by children
+
+# %%
+list(
```

Comment:
This community / group seems much better defined than the ones from the smaller dataset.